StitchCore/StitchCoreTests/TestHarness/TestHarness.swift

```diff
@@ -0,0 +1,172 @@
+                     serverUrl: String = defaultServerUrl) {
+        self.username = username
+        self.password = password
+        self.serverUrl = serverUrl
+        self.initPromise = Promise()
+        initPromise.then {
+            StitchAdminClientFactory.create(baseUrl: self.serverUrl)
+        }.done {
+            self.adminClient = $0
+        }.cauterize()
+    }
+
+    func teardown() -> Promise<Void> {
+        return self.app.remove().asVoid()
+    }
+
+    func authenticate() -> Promise<Void> {
+        return self.adminClient.authenticate(
+            provider: EmailPasswordAuthProvider.init(username: self.username,
+                                                     password: self.password)
+        ).then { (_) -> Promise<UserProfile> in
+            return self.adminClient.fetchUserProfile()
+        }.done { (userProfile: UserProfile) in
+            self.groupId = userProfile.roles!.first?.groupId
+        }
+    }
+
+    func createApp(testAppName: String = "test-\(ObjectId.init().hexString)") -> Promise<AppView> {
+        return self.apps.create(name: testAppName).flatMap {
+            self.testApp = $0
+            return $0
+        }
+    }
+
+    func createUser(email: String = "test_user@domain.com",
+                    password: String = "password") -> Promise<UserView> {
+        self.userCredentials = (username: email, password: password)
+        return self.app.users.create(data: UserCreator.init(email: email, password: password)).flatMap {
+            self.user = $0
+            return self.user
+        }
+    }
+
+    func add(serviceConfig: ServiceConfigs, withRules rules: Rule...) -> Promise<ServiceView> {
+        return self.app.services.create(data: serviceConfig).then { view in
+            return when(resolved: rules.map {
+                return self.app.services.service(withId: view.id).rules.create(data: $0)
+            }).flatMap { _ in view }
+        }
+    }
+    
+    func add(providerConfig: ProviderConfigs) -> Promise<AuthProviderView> {
+        return self.app.authProviders.create(data: providerConfig)
+    }
+
+    func addDefaultUserpassConfig() -> Promise<AuthProviderView> {
+        return self.add(providerConfig: .userpass(emailConfirmationUrl: "http://emailConfirmURL.com",
+                                                  resetPasswordUrl: "http://resetPasswordURL.com",
+                                                  confirmEmailSubject: "email subject",
+                                                  resetPasswordSubject: "password subject"))
+    }
+
+    func addDefaultAnonConfig() -> Promise<AuthProviderView> {
+        return self.add(providerConfig: .anon())
+    }
+
+    func addDefaultCustomTokenConfig() -> Promise<AuthProviderView> {
+        return self.add(providerConfig: .custom(signingKey: "abcdefghijklmnopqrstuvwxyz1234567890"))
+    }
+
+    func setupStitchClient(shouldConfigureUserAuth: Bool = true) -> Promise<Void> {
```

Comment:
[nit] param unnecessary or do we still want to implement the setup of an unauthed StitchClient?

Reply:
[q] on a similar note, shouldn't our linter or the swift compiler be catching these?

Reply:
Our linter isn't set up for the testing traget. That's a discussion we can have at the architecture meeting.